lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py

```diff
@@ -172,38 +296,107 @@ def ensure_map_public(conn, module, subnet, map_public, check_mode):
                          **camel_dict_to_snake_dict(e.response))
 
 
+def ensure_assign_ipv6_on_create(conn, module, subnet, assign_ipv6_on_create, check_mode):
+    if check_mode:
+        return
+
+    try:
+        conn.modify_subnet_attribute(SubnetId=subnet['id'], AssignIpv6AddressOnCreation={'Value': assign_ipv6_on_create})
+    except botocore.exceptions.ClientError as e:
+        module.fail_json(msg=e.message, exception=traceback.format_exc(),
+                         **camel_dict_to_snake_dict(e.response))
+
+
+def ensure_ipv6_cidr_block(conn, module, subnet, ipv6cidr, check_mode):
+    if check_mode:
+        return
+
+    if subnet['ipv6_association_id']:
+        # Make sure assign ipv6 on creation is turned off before disassociation
+        ensure_assign_ipv6_on_create(conn, module, subnet, False, check_mode)
+
+        try:
+            conn.disassociate_subnet_cidr_block(AssociationId=subnet['ipv6_association_id'])
+        except botocore.exceptions.ClientError as e:
+            # set assign ipv6 on creation value back to original value before failing
+            ensure_assign_ipv6_on_create(conn, module, subnet, subnet['assign_ipv6_address_on_creation'], check_mode)
```

Comment:
If both of these operations fail, it's going to be unclear what actually happened.

Reply:
I see what you are saying. The problem I have is that I have to turn off the `assign_ipv6_address_on_create` attribute before disassociating an IPv6 CIDR block and then set that attribute back to its intended value after assigning a new IPv6 CIDR. I was trying to avoid leaving the subnet in an unintended state on failure. i.e. `assign_ipv6` attribute gets disabled but then the disassociate operation fails for some reason.

I'm not sure if I, should be trying to correct from that failure, just provide more detailed error messages about the potentially bad state, move the `ensure_assign_ipv6_on_create` operation to the `try` block or to a `finally` block?